### PR TITLE
#185. Cluster configuration files. Initial step

### PR DESCRIPTION
**Pull request by sadovnikov**
Created for #185. Goal of this branch is to be able to start the cluster from configuration file, if present

### CODE REVIEW

File: minimesos/src/main/groovy/com/containersol/minimesos/config/ClusterConfig.groovy [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        cl.setResolveStrategy(Closure.DELEGATE_ONLY)
+        cl.call();
+    }
+
+    def exposePorts = true
+    int timeout = 60
+    String mesosVersion = "0.25"
+
+    def clusterName = "minimesos-test"
+
```
Comment (frankscholten) [inline minimesos/src/main/groovy/com/containersol/minimesos/config/ClusterConfig.groovy]: Perhaps call the cluster minimesos

File: minimesos/src/main/java/com/containersol/minimesos/main/CommandUp.java [hunk 2]
```diff
@@ -46,14 +44,24 @@ public String getMarathonImageTag() {
     private int timeout = MesosCluster.DEFAULT_TIMEOUT_SECS;
 
     @Parameter(names = "--num-agents", description = "Number of agents to start")
-    private int numAgents = 1;
+    private int numAgents = -1;
```
Comment (frankscholten) [inline minimesos/src/main/java/com/containersol/minimesos/main/CommandUp.java]: Why -1 agents?

File: minimesos/src/main/java/com/containersol/minimesos/main/CommandUp.java [hunk 3]
```diff
@@ -94,29 +125,110 @@ public void execute() {
...
 
-        startedCluster = new MesosCluster(configBuilder.build());
-        startedCluster.start(getTimeout());
-        startedCluster.waitForState(state -> state != null, 60);
-        startedCluster.setExposedHostPorts( isExposedHostPorts() );
-
-        startedCluster.printServiceUrls(output);
-
-        ClusterRepository.saveClusterFile(startedCluster);
+        return configBuilder;
```
Comment (frankscholten) [inline minimesos/src/main/java/com/containersol/minimesos/main/CommandUp.java]: I think this method should be in ClusterArchitecture

File: minimesos/src/main/java/com/containersol/minimesos/main/CommandUp.java [hunk 4]
```diff
@@ -94,29 +125,110 @@ public void execute() {
...
+        configBuilder = createCluster(configBuilder, clusterConfig);
 
-        ClusterArchitecture.Builder configBuilder = new ClusterArchitecture.Builder(dockerClient)
-                .withZooKeeper(getZooKeeperImageTag())
-                .withMaster(zooKeeper -> new MesosMasterExtended(dockerClient, zooKeeper, MesosMaster.MESOS_MASTER_IMAGE, getMesosImageTag(), new TreeMap<>(), isExposedHostPorts()))
-                .withContainer(zooKeeper -> new Marathon(dockerClient, zooKeeper, getMarathonImageTag(), isExposedHostPorts()), ClusterContainers.Filter.zooKeeper());
+        return configBuilder.build();
 
+    }
+
```
Comment (frankscholten) [inline minimesos/src/main/java/com/containersol/minimesos/main/CommandUp.java]: The above method should be in ClusterArchitecture

File: minimesos/src/main/java/com/containersol/minimesos/main/Main.java [hunk 5]
```diff
@@ -20,6 +21,9 @@
 
     private static final Logger LOGGER = LoggerFactory.getLogger(Main.class);
 
+    private static final int RC_OK = 0;
+    private static final int RC_ERR = 1;
+
```
Comment (frankscholten) [inline minimesos/src/main/java/com/containersol/minimesos/main/Main.java]: Change RC_OK and RC_ERR to more descriptive names

File: minimesos/src/main/java/com/containersol/minimesos/main/Main.java [hunk 6]
```diff
@@ -60,28 +72,30 @@ public void run(String[] args) {
...
                 printUsage(null);
+                return RC_ERR;
             }
-            return;
         }
 
         Command parsedCommand = commands.get(jc.getParsedCommand());
 
         if (parsedCommand == null) {
             LOGGER.error("No such command: " + jc.getParsedCommand());
```
Comment (frankscholten) [inline minimesos/src/main/java/com/containersol/minimesos/main/Main.java]: It seems jc.getParsedCommands() is checked for null twice.

File: minimesos/src/test/java/com/containersol/minimesos/main/CommandUpTest.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        CommandUp commandUp = new CommandUp();
+
+        ClusterArchitecture architecture = commandUp.getClusterArchitecture();
+        assertNotNull("architecture is not loaded", architecture);
+
+        ClusterContainers clusterContainers = architecture.getClusterContainers();
+        assertNotNull("cluster containers are not loaded", clusterContainers);
+
+        assertTrue("zooker is required component of cluster", clusterContainers.isPresent(ClusterContainers.Filter.zooKeeper()));
+        assertTrue("Mesos Master is required component of cluster", clusterContainers.isPresent(ClusterContainers.Filter.mesosMaster()));
```
Comment (frankscholten) [inline minimesos/src/test/java/com/containersol/minimesos/main/CommandUpTest.java]: Typo: 'zooker'

File: minimesos/src/main/java/com/containersol/minimesos/main/Main.java (hunk 5 above)
Comment (sadovnikov) [inline minimesos/src/main/java/com/containersol/minimesos/main/Main.java]: `RC_XXX` is not descriptive?  It stands for Return Code

File: minimesos/src/main/java/com/containersol/minimesos/main/Main.java (hunk 6 above)
Comment (sadovnikov) [inline minimesos/src/main/java/com/containersol/minimesos/main/Main.java]: No, first check if JCommander managed to read the command name. The second check if we have command Object matching the name

**Comment by sadovnikov** [comment]
@frankscholten Is anything open here? Are you OK with pressing the Merge button?

**Comment by sadovnikov** [comment]
@frankscholten this one is READY

**Comment by frankscholten** [comment]
LGTM

**Pull request merged**